### PR TITLE
Support filter clause in aggregations

### PR DESCRIPTION
**Pull request by chloe-zh**
*Issue #, if available:*
#964 

*Description of changes:*
- Added filter option for aggregate function in grammar
- Applied filter in aggregators
- Push down filter aggregation to ES DSL
- User doc: https://github.com/chloe-zh/sql/blob/adql/filter/docs/user/dql/aggregations.rst#filter-clause

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java [hunk 1]
```diff
@@ -62,28 +66,39 @@ public MetricAggregationBuilder(
...
         throw new IllegalStateException(
             String.format("unsupported aggregator %s", node.getFunctionName().getFunctionName()));
     }
   }
 
-  private ValuesSourceAggregationBuilder<?> make(ValuesSourceAggregationBuilder<?> builder,
-                                                  Expression expression) {
-    return helper.build(expression, builder::field, builder::script);
+  private QueryBuilder makeFilter(Expression condition) {
+    return new FilterQueryBuilder(serializer).build(condition);
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java:91]: could we make FilterQueryBuilder as depedency?

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java [hunk 2]
```diff
@@ -62,28 +66,39 @@ public MetricAggregationBuilder(
...
+  }
+
+  private AggregationBuilder make(ValuesSourceAggregationBuilder<?> builder,
+                                  Expression expression, Expression condition, String name) {
+    ValuesSourceAggregationBuilder aggBuilder =
+        helper.build(expression, builder::field, builder::script);
+    if (condition == null) {
+      return aggBuilder;
+    }
+    return AggregationBuilders.filter(name, makeFilter(condition)).subAggregation(aggBuilder);
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java:101]: could we move these logic as a decoration to hide the implementation details, e.g.
```
return filter(childAggBuilder, name, condition)
```

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/aggregation/AvgAggregator.java [hunk 3]
```diff
@@ -46,7 +46,7 @@ public AvgState create() {
   public AvgState iterate(BindingTuple tuple, AvgState state) {
     Expression expression = getArguments().get(0);
     ExprValue value = expression.valueOf(tuple);
-    if (!(value.isNull() || value.isMissing())) {
+    if (!(value.isNull() || value.isMissing()) && conditionValue(tuple)) {
```
Comment (dai-chen) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/aggregation/AvgAggregator.java:49]: Any chance to move this template code in each subclass to the parent class `Aggregator`? As what I see, each aggregator impl class only differs on how to accumulate state when this condition is true.

**Review by dai-chen — COMMENTED** [review]
Please create an issue for this new feature. Thanks!

File: integ-test/src/test/resources/correctness/queries/filter.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+SELECT AVG(AvgTicketPrice) FILTER(WHERE Carrier = 'Kibana Airlines') AS filtered FROM kibana_sample_data_flights
```
Comment (dai-chen) [inline integ-test/src/test/resources/correctness/queries/filter.txt:1]: Could you add test case to check if we can support function call in `WHERE`? Just want to check if pushdown works in this case.

**Review by dai-chen — APPROVED** [review]
LGTM. Thanks!

**Review by penghuo — APPROVED** [review]

**Pull request merged**